language/stdlib/modules/VASP.move

```diff
@@ -174,6 +180,138 @@ module VASP {
         borrow_global_mut<ParentVASP>(parent_addr).compliance_public_key = new_key
     }
 
+    // **************** SPECIFICATIONS ****************
+
+    /// # Module specifications
+
+    spec module {
+        /// > TODO(emmazzz): verification is turned off because `Signature` module
+        /// > has not been specified. See issue #4666.
+        pragma verify = false;
+    }
+
+    spec module {
+        /// Returns true if `addr` is a VASP.
+        define spec_is_vasp(addr: address): bool {
+            exists<ParentVASP>(addr) || exists<ChildVASP>(addr)
+        }
+
+        /// Returns true if `addr` is a ParentVASP.
+        define spec_is_parent_vasp(addr: address): bool {
+            exists<ParentVASP>(addr)
+        }
+
+        /// Returns true if `addr` is a ChildVASP.
+        define spec_is_child_vasp(addr: address): bool {
+            exists<ChildVASP>(addr)
+        }
+
+        /// Returns the number of children under `parent`.
+        define spec_get_num_children(parent: address): u64 {
+            global<ParentVASP>(parent).num_children
+        }
+
+        /// Returns the parent address of a VASP.
+        define spec_parent_address(addr: address): address {
+            if (exists<ParentVASP>(addr)) {
+                addr
+            } else {
+                global<ChildVASP>(addr).parent_vasp_addr
+            }
+        }
+
+        define spec_cert_lifetime(): u64 {
+            31540000000000
+        }
+
+        define spec_root_address(): address {
+            0xA550C18
+        }
+    }
+
+
+    /// ## Number of children is consistent
+    /// > PROVER TODO(emmazzz): implement the features that allows users
+    /// > to reason about number of resources with certain property,
+    /// > such as "number of ChildVASPs whose parent address is 0xDD".
+    /// > See issue #4665.
+
+
+    /// ## Number of children does not change
+
+    spec schema NumChildrenRemainsSame {
+        ensures forall parent: address
+            where old(spec_is_parent_vasp(parent)):
+                old(spec_get_num_children(parent))
+                 == spec_get_num_children(parent);
+    }
+
+    spec module {
+        apply NumChildrenRemainsSame to * except publish_child_vasp_credential;
+    }
+
+    /// ## Specifications for individual functions
+    spec schema AbortsIfNotVASP {
+        addr: address;
+        aborts_if !spec_is_vasp(addr);
+    }
+
+    spec module {
```

Comment:
This is an interesting style.  I think Wolfgang used "include" in spec fun <whatever>  to reduce repetition.  This is kind of like the transpose of that (specs -> functions, not functions -> specs).  I don't know which is better.